### PR TITLE
Add `useBlockProtocolGetEntity` hook + example entity-editor page

### PR DESCRIPTION
**Pull request by benwerner01**
## 🌟 What is the purpose of this PR?

This PR adds the `useBlockProtocolGetEntity` hook to be used for the upcoming entity editor.

It also provides an example entity editor page to the frontend, demonstrating its usage.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1202805690238892/1203084714149805/f) _(internal)_

## 🔍 What does this change?

See commits.

## 📜 Does this require a change to the docs?

No.

## 🛡 What tests cover this?

None.

## ❓ How to test this?

1. Checkout this branch
2. Run `yarn external-services up` locally
3. Run `yarn dev`
4. Open `localhost:3000` and login
5. Go to `localhost:3000/entity-editor` to view the example entity editor making use of the `useBlockProtocolGetEntity` method

## 📹 Demo

https://user-images.githubusercontent.com/42802102/194020333-bd8f0368-aa57-43f9-953b-6e9494978e30.mov

### CODE REVIEW

File: packages/hash/frontend/src/components/hooks/blockProtocolFunctions/ontology/useBlockProtocolGetDataType.ts [hunk 1]
```diff
@@ -32,8 +32,11 @@ export const useBlockProtocolGetDataType = (): {
         };
       }
 
+      const { dataTypeId } = data;
+
       const response = await getFn({
         query: getDataTypeQuery,
+        variables: { dataTypeId },
       });
```
Comment (benwerner01) [inline packages/hash/frontend/src/components/hooks/blockProtocolFunctions/ontology/useBlockProtocolGetDataType.ts:40]: These are drive-by fixes for the existing `useBlockProtocolGetDataType`/`EntityType`/`PropertyType`/`LinkType` hooks, which to my understanding don't work in the current implementation because they aren't passing required variables to their respective queries

File: packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/knowledge-shim.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+import { MessageCallback } from "@blockprotocol/core";
+import { ReadOrModifyResourceError } from "@blockprotocol/graph";
+import {
+  UnknownKnowledgeEntity,
+  EntityTypeRootedSubgraph,
+} from "../../../../graphql/apiTypes.gen";
+
+export type KnolwedgeCallbacks = {
```
Comment (Alfred-Mountfield) [inline packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/knowledge-shim.ts:16]:
```suggestion
export type KnowledgeCallbacks = {
```

File: packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolGetEntity.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+import { GetEntityMessageCallback } from "./knowledge-shim";
+
+export const useBlockProtocolGetEntity = (): {
+  getEntity: GetEntityMessageCallback;
+} => {
+  const [getEntityTypeRootedSubgraphFn] = useLazyQuery<
+    GetEntityTypeRootedSubgraphQuery,
+    GetEntityTypeRootedSubgraphQueryVariables
+  >(getEntityTypeRootedSubgraphQuery, {
+    /** @todo reconsider caching. This is done for testing/demo purposes. */
```
Comment (Alfred-Mountfield) [inline packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolGetEntity.ts:21]: When do you want to address this, before merge?

File: packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolGetEntity.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+              message: "Error calling getEntity",
+            },
+          ],
+        };
+      }
+
+      const { knowledgeEntity } = entityResponseData;
+
+      const { data: entityTypeResponseData } =
+        await getEntityTypeRootedSubgraphFn({
```
Comment (Alfred-Mountfield) [inline packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolGetEntity.ts:67]: Wonder if it's worth a TODO to say that these two should be one query to get an EntityRootedSubgraph

File: packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolGetEntity.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+          },
+        });
+
+      if (!entityTypeResponseData) {
+        return {
+          errors: [
+            {
+              code: "INVALID_INPUT",
+              message:
+                "Error getting the entity type rooted sub graph of the entity",
```
Comment (Alfred-Mountfield) [inline packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolGetEntity.ts:80]:
```suggestion
                "Error getting the entity type rooted subgraph of the entity",
```

or maybe
```suggestion
                "Error getting the subgraph rooted at entity's entity type",
```

File: packages/hash/frontend/src/pages/entity-editor.page.tsx [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+ */
+const ExampleUsage = ({ ownedById }: { ownedById: string }) => {
+  const { user } = useUser();
+  const [entity, setEntity] = useState<EntityResponse>();
+
+  const { getEntity } = useBlockProtocolFunctionsWithOntology(ownedById);
+
+  useEffect(() => {
+    if (user) {
+      // As an example entity, we are going to use the currenlty logged in user's entity id
```
Comment (Alfred-Mountfield) [inline packages/hash/frontend/src/pages/entity-editor.page.tsx:30]:
```suggestion
      // As an example entity, we are going to use the currently logged in user's entity ID
```

File: packages/hash/frontend/src/pages/entity-editor.page.tsx [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        setEntity(data);
+      });
+    }
+  }, [user, getEntity]);
+
+  const { entityTypeRootedSubgraph, ...entityWithoutEntityType } = entity ?? {};
+
+  const { entityType } = entityTypeRootedSubgraph ?? {};
+
+  // The (top-level) property type ids defined in the entity type
```
Comment (Alfred-Mountfield) [inline packages/hash/frontend/src/pages/entity-editor.page.tsx:43]:
```suggestion
  // The (top-level) property type IDs defined in the entity type
```

File: packages/hash/frontend/src/pages/type-editor/blockprotocol-ontology-functions-hook.ts [hunk 8]
```diff
@@ -40,6 +40,8 @@ import { useBlockProtocolGetLinkType } from "../../components/hooks/blockProtoco
 import { useBlockProtocolUpdateLinkType } from "../../components/hooks/blockProtocolFunctions/ontology/useBlockProtocolUpdateLinkType";
 
 import { useReadonlyMode } from "../../shared/readonly-mode";
+import { useBlockProtocolGetEntity } from "../../components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolGetEntity";
+import { KnolwedgeCallbacks } from "../../components/hooks/blockProtocolFunctions/knowledge/knowledge-shim";
```
Comment (Alfred-Mountfield) [inline packages/hash/frontend/src/pages/type-editor/blockprotocol-ontology-functions-hook.ts:44]:
```suggestion
import { KnowledgeCallbacks } from "../../components/hooks/blockProtocolFunctions/knowledge/knowledge-shim";
```

File: packages/hash/frontend/src/pages/type-editor/blockprotocol-ontology-functions-hook.ts [hunk 9]
```diff
@@ -54,7 +56,8 @@ export type GraphMessageCallbacks = Omit<
   | "updateEntityType"
   | "aggregateEntityTypes"
 > &
-  OntologyCallbacks;
+  OntologyCallbacks &
+  KnolwedgeCallbacks;
```
Comment (Alfred-Mountfield) [inline packages/hash/frontend/src/pages/type-editor/blockprotocol-ontology-functions-hook.ts:60]:
```suggestion
  KnowledgeCallbacks;
```

**Review by Alfred-Mountfield — CHANGES_REQUESTED** [review]
Looks good to me for the most part. A typo and some nits on phrasing

File: packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolGetEntity.ts (hunk 4 above)
Comment (benwerner01) [reply packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolGetEntity.ts:67]: Good idea, is there a task for the upcoming GQL refactor work?

Comment (Alfred-Mountfield) [reply packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolGetEntity.ts:67]: Probably going to end up encapsulate in [this task](https://app.asana.com/0/1200211978612931/1203089535761796/f) (might rework its title and description)

File: packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolGetEntity.ts (hunk 3 above)
Comment (benwerner01) [reply packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolGetEntity.ts:21]: This is a copy and paste from the other BP methods, I don't think we should address before merging no.

Comment (benwerner01) [reply packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolGetEntity.ts:21]: @thehabbos007 is there a task for this?

Comment (Alfred-Mountfield) [reply packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolGetEntity.ts:21]: Makes sense, thanks

Comment (Alfred-Mountfield) [reply packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolGetEntity.ts:21]: @thehabbos007 do we want an asana task to track this? Should it be part of the clean-up post merge?

Comment (thehabbos007) [reply packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolGetEntity.ts:21]: No task for this yet, i think it's good to have it be `no-cache` for a while, and we might only want to allow cached versions if the entity version IDs are passed in the query as well. I think caching could end up causing a lot of headache for frontend development, so best to leave it out for now.

We can create a task, though

File: packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolGetEntity.ts (hunk 4 above)
Comment (benwerner01) [reply packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolGetEntity.ts:72]: Added the todo in https://github.com/hashintel/hash/pull/1160/commits/03b165ace4db95f35a3c55c7b788470c9413087b

File: packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolGetEntity.ts (hunk 3 above)
Comment (Alfred-Mountfield) [reply packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolGetEntity.ts:21]: It's probably fine to leave without a task. We'll encounter this when we remove the shims and have actual BP methods

**Review by Alfred-Mountfield — APPROVED** [review]
Provided my suggested changes didn't break the code, this looks good to me 🚀

File: packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolGetEntity.ts (hunk 4 above)
Comment (benwerner01) [reply packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolGetEntity.ts:72]: Added the todo in https://github.com/hashintel/hash/pull/1160/commits/03b165ace4db95f35a3c55c7b788470c9413087b

File: packages/hash/frontend/src/graphql/queries/ontology/entity-type.queries.ts [hunk 10]
```diff
@@ -20,6 +20,47 @@ export const getAllLatestEntityTypesQuery = gql`
...
+  ) {
+    getEntityType(entityTypeId: $entityTypeId) {
+      entityTypeId
+      ownedById
+      accountId
+      entityType
+      referencedDataTypes(depth: $referencedDataTypesDepth) {
+        dataTypeId
+        ownedById
+        accountId
```
Comment (thehabbos007) [inline packages/hash/frontend/src/graphql/queries/ontology/entity-type.queries.ts:39]: Might have missed this somewhere, but how come we have the `accountId` in this query well when we control the ontology types?

Comment (benwerner01) [reply packages/hash/frontend/src/graphql/queries/ontology/entity-type.queries.ts:39]: Good point, this can be removed I think but will have to be done in follow-up. Will do a drive-by in an upcoming PR
